nni/algorithms/compression/pytorch/pruning/iterative_pruner.py

```diff
@@ -0,0 +1,574 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import logging
+import copy
+import torch
+from schema import And, Optional, SchemaError
+from nni.compression.pytorch.utils.config_validation import CompressorSchema
+from nni.compression.pytorch.compressor import Pruner
+from .constants import MASKER_DICT
+from .dependency_aware_pruner import DependencyAwarePruner
+
+__all__ = ['AGPPruner', 'ADMMPruner', 'SlimPruner', 'TaylorFOWeightFilterPruner', 'ActivationAPoZRankFilterPruner', 'ActivationMeanRankFilterPruner', ]
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class IterativePruner(DependencyAwarePruner):
+    """
+    Prune model during the training process. 
+    """
+
+    def __init__(self, model, config_list, optimizer=None, pruning_algorithm='level', trainer=None, criterion=None, training_epochs=20, dependency_aware=False, dummy_input=None, **algo_kwargs):
+        """
+        Parameters
+        ----------
+        model : torch.nn.Module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        optimizer: torch.optim.Optimizer
+            Optimizer used to train model
+        pruning_algorithm: str
+            algorithms being used to prune model
+        trainer: function
+            Function used to train the model.
+            Users should write this function as a normal function to train the Pytorch model
+            and include `model, optimizer, criterion, epoch, callback` as function arguments.
+        criterion: function
+            Function used to calculate the loss between the target and the output.
+        training_epochs : int
+            Totoal number of training epochs for iterative pruning.
+        dependency_aware: bool
+            If prune the model in a dependency-aware way.
+        dummy_input : torch.Tensor
+            The dummy input to analyze the topology constraints. Note that, the dummy_input should on the same device with the model.
+        algo_kwargs: dict
+            Additional parameters passed to pruning algorithm masker class
+        """
+        super().__init__(model, config_list, optimizer, pruning_algorithm, dependency_aware, dummy_input, **algo_kwargs)
+
+        self.training_epochs = training_epochs
+        self._trainer = trainer
+        self._criterion = criterion
+        
+
+    def compress(self):
+        training = self.bound_model.training
+        self.bound_model.train()
+        for epoch in range(self.training_epochs):
+            self._trainer(self.bound_model, optimizer=self.optimizer,
+            criterion=self._criterion, epoch=epoch, callback=self._callback)
+        self.bound_model.train(training)
+
+        self.update_mask()
```

Comment:
is this `compress` function iterative? seems `update_mask` is put outside of the for loop

Reply:
is our SlimPruner not precisely implemented?